### DISCUSSION TITLE
Extend the Bot Universe to include hierarchies

### DISCUSSION THREAD

**Discussion by HaraldHuber01** [USUBotUniverse/USUBotUniverse#5] [Ideas]
We have a lot of experience today with bot universes with up to > 10 bot experts. In some projects we have received a request that it would make sense to structure the experts into "sub-universes" - or let's call them galaxies or whatever.
In fact, the idea could be to organise the experts in such a way that several experts work together on a particular topic.
For example, a bot universe could contain several experts working on the Knowledge Center theme (which is our knowledge base software). There could be one expert for error messages, one for the new active documents and one for the e-learning module...... These experts can form a "sub-universe".
Maybe someone can help me with this idea?

**Comment by usimschleg** [comment]
Hmm, an interesting idea.

This would allow bots to be even more optimised for their special field and thus provide higher-quality answers. But does this also mean that a bot galaxy has its own black hole ;-) (lead bot) through which the accesses are distributed?

**Reply by HaraldHuber01** [reply]
Yes, in order to fulfill the "black box" idea every galaxy has to have an own Leadbot - otherwise a galaxy would behave different than an expert.

**Comment by HenryUSU** [comment]
I like the idea.
However, we have to prevent that loops can be created by the user. 
E.g. this should not be possible:
![grafik](https://user-images.githubusercontent.com/97042465/147950951-bca8ac11-b268-47e8-9d35-8d58a264172f.png)

**Reply by HaraldHuber01** [reply]
Hello Henry, 

yes, I think that makes sense. 
From my perspective, it's not the big risk. 
Maybe we can first implement an extension to the behavior of lead bots, so that a hierarchy of bot universes is possible. And then we could implement a "listing feature" so that a chatbot creates a list of all universes it belongs to. Also transitive. Then if a universe is listed twice, there is a loop. 
But as a first step, the architect has to make sure that this kind of thing is avoided.

**Comment by HaraldHuber01** [comment]
One comment: 
I just discussed the issue with the customer GISA. 
They would appreciate the feature. It would really fit their needs. Today they have a bot universe with an expert "Applications". They want to extend that so that "applications" can again be a universe built from "office", "sap", etc. 
So also from their point of view it makes sense
